// action-package-deploy/src/action-auth.js
const path = require("path");
const express = require('express');
const opn = require('opn');
const uuid = require('uuid');

const client_id = "cac88df7-3599-49cf-9465-867b9eee33cf";
const scope = "https://actions.office365.com/ActionPackage.ReadWrite.All";
const port = 22222;
const redirect_uri = `http://localhost:${port}/ActionsPlatform`;

function handleTokenResponse(resolve, reject, tokenResponse, state, res, server) {
    // Ignore if we haven't requested this token response
    if (tokenResponse.state != state) {
        return;
    }

    // Handle the token response
    var resposeMessage;
    if (tokenResponse.error || tokenResponse.error_description) {
        // Login failed
        resposeMessage = "Login failed! Please close this window";
        reject({
            errorCode: tokenResponse.error,
            errorMessage: tokenResponse.error_description || tokenResponse.error_subcode
        });
    } else {
        // Login succeeded
        resposeMessage = "Login successful! Please close this window";
        resolve({
            token: tokenResponse.access_token,
            expiresIn: tokenResponse.expires_in
        });
    }

    // Send response
    res.send(resposeMessage);

    // Close the server
    server.close();
}

exports.fetchAuthToken = async () => {
    return new Promise((resolve, reject) => {
        var state = uuid.v4();
        var nonce = uuid.v4();

        var app = express();
        var server;

        app.get('/ActionsPlatform', function (req, res) {
            // Login is complete, open auth page to transfer us the token
            var post_login_page = path.resolve(__dirname, "action-auth-page.html");
            res.sendFile(post_login_page);
        });

        app.get('/AuthToken', function (req, res) {
            handleTokenResponse(resolve, reject, req.query, state, res, server);
        });

        // Start the server
        server = app.listen(port);

        // opens the login url in default browser
        var authUrl = `https://login.microsoftonline.com/common/oauth2/v2.0/authorize?response_type=token&response_mode=fragment&prompt=select_account&client_id=${client_id}&redirect_uri=${redirect_uri}&scope=${scope}&state=${state}&nonce=${nonce}`;
        opn(authUrl);
    });
}
